core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -186,30 +187,14 @@ private Object convertMapsToArrays(Map<String, Object> map) {
      * A settings that are filtered (and key is removed) with the specified prefix.
      */
     public Settings getByPrefix(String prefix) {
-        Builder builder = new Builder();
-        for (Map.Entry<String, String> entry : getAsMap().entrySet()) {
-            if (entry.getKey().startsWith(prefix)) {
-                if (entry.getKey().length() < prefix.length()) {
-                    // ignore this. one
-                    continue;
-                }
-                builder.put(entry.getKey().substring(prefix.length()), entry.getValue());
-            }
-        }
-        return builder.build();
+        return new Settings(new FilteredMap(this.settings, (k) -> k.startsWith(prefix) && k.length() >= prefix.length(), prefix));
```

Comment:
I think `k.length() >= prefix.length()` is always true when `k.startsWith(prefix)`?